src/array.rs

```diff
@@ -0,0 +1,456 @@
+}
+
+macro_rules! impl_array_element {
+    ($type:ty, $dtype:expr, $mlx_item_fn:ident, $mlx_data_fn:ident) => {
+        impl Sealed for $type {}
+        impl ArrayElement for $type {
+            const DTYPE: Dtype = $dtype;
+
+            fn scalar_array_item(array: &Array) -> Self {
+                unsafe { mlx_sys::$mlx_item_fn(array.ptr) }
+            }
+
+            fn array_data(array: &Array) -> *const Self {
+                unsafe { mlx_sys::$mlx_data_fn(array.ptr) }
+            }
+        }
+    };
+}
+
+impl_array_element!(bool, Dtype::Bool, mlx_array_item_bool, mlx_array_data_bool);
+impl_array_element!(u8, Dtype::Uint8, mlx_array_item_uint8, mlx_array_data_uint8);
+impl_array_element!(
+    u16,
+    Dtype::Uint16,
+    mlx_array_item_uint16,
+    mlx_array_data_uint16
+);
+impl_array_element!(
+    u32,
+    Dtype::Uint32,
+    mlx_array_item_uint32,
+    mlx_array_data_uint32
+);
+impl_array_element!(
+    u64,
+    Dtype::Uint64,
+    mlx_array_item_uint64,
+    mlx_array_data_uint64
+);
+impl_array_element!(i8, Dtype::Int8, mlx_array_item_int8, mlx_array_data_int8);
+impl_array_element!(
+    i16,
+    Dtype::Int16,
+    mlx_array_item_int16,
+    mlx_array_data_int16
+);
+impl_array_element!(
+    i32,
+    Dtype::Int32,
+    mlx_array_item_int32,
+    mlx_array_data_int32
+);
+impl_array_element!(
+    i64,
+    Dtype::Int64,
+    mlx_array_item_int64,
+    mlx_array_data_int64
+);
+impl_array_element!(
+    f32,
+    Dtype::Float32,
+    mlx_array_item_float32,
+    mlx_array_data_float32
+);
+
+impl Sealed for f16 {}
+
+impl ArrayElement for f16 {
+    const DTYPE: Dtype = Dtype::Float16;
+
+    fn scalar_array_item(array: &Array) -> Self {
+        let val = unsafe { mlx_sys::mlx_array_item_float16(array.ptr) };
+        f16::from_bits(val.0)
+    }
+
+    fn array_data(array: &Array) -> *const Self {
+        unsafe { mlx_sys::mlx_array_data_float16(array.ptr) as *const Self }
+    }
+}
+
+impl Sealed for bf16 {}
+
+impl ArrayElement for bf16 {
+    const DTYPE: Dtype = Dtype::Bfloat16;
+
+    fn scalar_array_item(array: &Array) -> Self {
+        let val = unsafe { mlx_sys::mlx_array_item_bfloat16(array.ptr) };
+        bf16::from_bits(val)
+    }
+
+    fn array_data(array: &Array) -> *const Self {
+        unsafe { mlx_sys::mlx_array_data_bfloat16(array.ptr) as *const Self }
+    }
+}
+
+impl Sealed for c64 {}
+
+impl ArrayElement for c64 {
+    const DTYPE: Dtype = Dtype::Complex64;
+
+    fn scalar_array_item(array: &Array) -> Self {
+        let bindgen_c64 = unsafe { mlx_sys::mlx_array_item_complex64(array.ptr) };
+
+        Self {
+            re: bindgen_c64.re,
+            im: bindgen_c64.im,
+        }
+    }
+
+    fn array_data(array: &Array) -> *const Self {
+        // c64 has the same memory layout as __BindgenComplex<f32>
+        unsafe { mlx_sys::mlx_array_data_complex64(array.ptr) as *const Self }
+    }
+}
+
+pub struct Array {
+    ptr: mlx_array,
```

Comment:
That would work, but we'll need this for almost every other type. i.e. device, stream, etc. Having just one name makes it easier rather than specifying for each.  We could just prefix with `c_` though: `c_array`: `c_device` `c_stream`. let's do that.

Reply:
Sounds good